### DISCUSSION TITLE
Request Logging Customization

### DISCUSSION THREAD

**Discussion by kdrewes** [fastify/fastify#5726] [Q&A]
Hello, I have a question regarding my last PR I attempted to merge:

https://github.com/fastify/fastify/pull/4955

this PR originally derived from the PR below:

https://github.com/fastify/fastify/issues/4463

The main objective of this task is to build a feature which will allow the user/client the ability to produce a customizable message within the log management system.  Within this message, the user/client will have the capability to configure how they want the message to be constructed.  

As of now I've attempted to follow the recommendations provided to me however I've been struggling to understand exactly how this feature is supposed to be constructed and wanted to see if someone had any advice or suggestions on how to proceed.  

Here's my current understanding of the situation at hand:
As of now, the log management system is set up in fastify.js as follows:
`
  if (disableRequestLogging === false) {
            childLogger.info({ req: request }, 'incoming request')
          }
`

As a result, when a request is made, fastify.js will be the first file triggered.  From there, once the condition is read, the log management system will automatically be executed, allow the user/client to view information and details pertaining to the request.  However, there is currently no feature which allows the user to customize this information.  Therefore, it was suggested in the original PR #4463 that we create a function, _createRequestLogMessage_, which will allow the user/client to customize the properties of the log management system.  

So far, I've attempted to declare this function as a prototype on fastify.d.ts.  

createRequestLogMessage?: (childLoggerOptions: ChildLoggerOptions, fastifyRequest: FastifyRequest) => string;

Then in fastify.js

I declared the createRequestLogMessage function as follows:

_CreateRequestLogMessage = options.createRequestLogMessage_

From there, I modified the following condition 

`
  if (disableRequestLogging === false) {
            childLogger.info({ req: request }, 'incoming request')
          }
`

to

`
if (!disableRequestLogging) {
        const logMessage = createRequestLogMessage(req)
        if (logMessage) {
          childLogger.info(childLogger, ...logMessage)
        } else {
          childLogger.info(childLogger, [{ req: request }, 'incoming request'])
        }
      }
`

Originally, this syntax worked, but I was requested by the maintainers to turn createRequestLogMessage into a default function and return the value accordingly as follows:

`
childLogger.info.call(childLogger, createRequestLogMessage(req))

`

Originally, adding 'call' did not work for me, so after some experimenting I came up with the following:

Built my own createRequestLogMessge (req)

`
  function createRequestLogMessage (req) {
    // Return incoming request prompt
    return [{ req }, 'incoming request']
  }

`
and then within the onBadUrl function I added the following:

`
 const logMessage = createRequestLogMessage(req)
      childLogger.info(childLogger, ...logMessage)
`

After running _npm run _test_ I received the following errors:

![image](https://github.com/user-attachments/assets/70a9cc24-4ed3-47cf-ba5e-97020e4bac04)
![image](https://github.com/user-attachments/assets/390252da-0ff3-4bf5-8cc6-a69ee874cec3)
![image](https://github.com/user-attachments/assets/a0ceff06-20ee-4125-96a9-74bb0472a86b)
![image](https://github.com/user-attachments/assets/98584495-3926-468b-bb5c-04efee3b4131)

So from reading the description of the errors, here's my hypothesis of the problem:

• Errors may be occurring due to the absence of disableRequestLogging condition.  Even though I was still encountering error when I attempted to use it.
• The following syntax is failing the test because there is an additional plug-in that needs to be included.
• Perhaps the function should be implemented in a different way which is causing an exception to be thrown?

Please let me know if you have any questions or critiques about this concern.  Thank you.

Kyle

**Comment by Eomm** [comment]
It is easier to talk on a PR with the code and the CI results instead of reasoning on a discussion.

The `childLogger.info.call(childLogger, createRequestLogMessage(req))` suggestion is needed to avoid a performance drop. Given that the log code is executed for each request, it means that if the server gets 1000 req/sec it will run a spread operator 1000 time/sec. This can have a notable performance drop. At work, I had one recent perf improvement of 600% by removing a spread operator.

So, given that the PR is small, neverthless it is editing a **core part** of the fastify framework.

If some tests are failing we should isolate them using the [`only` option](https://node-tap.org/plugins/filter/#only) and understand WHY the test is failing.
Each test can be checked individually to understand if:

- is the test correct or should we fix the test? Maybe we do have a different log pattern in the BAD_URL that should be updated or fixed. IDK and checking it right now, would mean implementing the feature TBH 😄 
- is the test valid and are we missing an edge case?

> Perhaps the function should be implemented in a different way which is causing an exception to be thrown?

It is one of the greatest question in the backend world: if a log message is broken, should my server crash? If a user implements a custom logger serializer ATM, it should fail if I'm not wrong, so I would keep the same behaviour.

**Reply by kdrewes** [reply]
Hi @Eomm, thank you so much for your feedback.  I will attempt to use the [only option](https://node-tap.org/plugins/filter/#only) for each test and evaluate the issue.  From there, I will get back to you and let you know what I discover.  Also, should I continue this conversation through the PR:

https://github.com/fastify/fastify/pull/4955

Thank you for all your help.

Kyle
